File: rust/acquisition-worker/src/quality.rs
use crate::models::QualityReport;
use regex::Regex;
use std::collections::HashMap;

pub struct QualityEvaluator {
    block_patterns: Vec<Regex>,
    spa_patterns: Vec<Regex>,
}

impl Default for QualityEvaluator {
    fn default() -> Self {
        Self::new()
    }
}

impl QualityEvaluator {
    pub fn new() -> Self {
        let block_patterns = vec![
            Regex::new(r"(?i)cloudflare").unwrap(),
            Regex::new(r"(?i)attention required").unwrap(),
            Regex::new(r"(?i)access denied").unwrap(),
            Regex::new(r"(?i)captcha").unwrap(),
            Regex::new(r"(?i)datadome").unwrap(),
            Regex::new(r"(?i)perimeterx").unwrap(),
            Regex::new(r"(?i)security check").unwrap(),
            Regex::new(r"(?i)please verify you are a human").unwrap(),
            Regex::new(r"(?i)bot detection").unwrap(),
            Regex::new(r"(?i)checking your browser before accessing").unwrap(),
        ];

        let spa_patterns = vec![
            Regex::new(r#"(?i)<div\s+id=["']root["']\s*>\s*</div>"#).unwrap(),
            Regex::new(r#"(?i)<div\s+id=["']app["']\s*>\s*</div>"#).unwrap(),
            Regex::new(r#"(?i)<div\s+id=["']__next["']\s*>\s*</div>"#).unwrap(),
            Regex::new(r"(?i)you need to enable javascript to run this app").unwrap(),
            Regex::new(r"(?i)please enable javascript").unwrap(),
        ];

        Self {
            block_patterns,
            spa_patterns,
        }
    }

    pub fn evaluate(
        &self,
        _url: &str,
        status_code: u16,
        _headers: &HashMap<String, String>,
        html_or_text: &str,
        expected_min_text_chars: usize,
    ) -> QualityReport {
        let mut reasons = Vec::new();
        let mut score: f64 = 1.0;
        let mut completeness: f64 = 1.0;
        let mut blocked = false;
        let mut likely_unrendered = false;
        let mut suggested_escalation = None;

        // 1. HTTP Status code
        if status_code >= 400 {
            score -= 0.6;
            completeness = 0.0;
            reasons.push(format!("HTTP error status: {}", status_code));
            if status_code == 403 || status_code == 429 {
                blocked = true;
                reasons.push("Rate limit or access forbidden".to_string());
                suggested_escalation = Some("chromium".to_string());
            }
        }

        // 2. Block patterns
        let sample_len = html_or_text.len().min(10000);
        let sample = &html_or_text[..sample_len];

        for pattern in &self.block_patterns {
            if pattern.is_match(sample) {
                blocked = true;
                score -= 0.5;
                reasons.push(format!("Block pattern detected: {}", pattern.as_str()));
                suggested_escalation = Some("chromium".to_string());
                break;
            }
        }

        // 3. Unrendered SPA shell detection
        for pattern in &self.spa_patterns {
            if pattern.is_match(sample) {
                likely_unrendered = true;
                score -= 0.4;
                completeness = completeness.min(0.3);
                reasons.push("Empty SPA root or disabled JS message detected".to_string());
                if suggested_escalation.is_none() {
                    suggested_escalation = Some("servo".to_string());
                }
                break;
            }
        }

        // 4. Useful text characters estimation
        let tag_regex = Regex::new(r"<[^>]+>").unwrap();
        let stripped = tag_regex.replace_all(html_or_text, " ");
        let text_chars = stripped.split_whitespace().map(|w| w.len()).sum::<usize>();

        if text_chars < expected_min_text_chars {
            score -= 0.3;
            let ratio = (text_chars as f64) / (expected_min_text_chars.max(1) as f64);
            completeness = completeness.min(ratio.max(0.1));
            reasons.push(format!(
                "Low useful text characters: {} < {}",
                text_chars, expected_min_text_chars
            ));
            if suggested_escalation.is_none() && !blocked {
                suggested_escalation = Some("servo".to_string());
            }
        }

        QualityReport {
            score: (score.clamp(0.0, 1.0) * 1000.0).round() / 1000.0,
            completeness: (completeness.clamp(0.0, 1.0) * 1000.0).round() / 1000.0,
            blocked,
            likely_unrendered,
            reasons,
            suggested_escalation,
        }
    }
}

File: scraper/acquisition/quality.py
"""Acquisition Quality Evaluator (§4, DS-RB05).

Evaluates page quality beyond HTTP 200, detecting empty SPA shells, false-200 block pages,
and suggesting optimal backend escalation paths.
"""

import re
from typing import Dict, List, Optional
from scraper.acquisition.models import QualityReport


class AcquisitionQualityEvaluator:
    """Computes comprehensive quality score and escalation hints."""

    BLOCK_PATTERNS = [
        re.compile(r"cloudflare", re.I),
        re.compile(r"attention required", re.I),
        re.compile(r"access denied", re.I),
        re.compile(r"captcha", re.I),
        re.compile(r"datadome", re.I),
        re.compile(r"perimeterx", re.I),
        re.compile(r"security check", re.I),
        re.compile(r"please verify you are a human", re.I),
        re.compile(r"bot detection", re.I),
        re.compile(r"checking your browser before accessing", re.I),
    ]

    UNRENDERED_SPA_PATTERNS = [
        re.compile(r"<div\s+id=[\"']root[\"']\s*>\s*</div>", re.I),
        re.compile(r"<div\s+id=[\"']app[\"']\s*>\s*</div>", re.I),
        re.compile(r"<div\s+id=[\"']__next[\"']\s*>\s*</div>", re.I),
        re.compile(r"you need to enable javascript to run this app", re.I),
        re.compile(r"please enable javascript", re.I),
    ]

    def evaluate(
        self,
        url: str,
        status_code: int,
        headers: Dict[str, str],
        html_or_text: str,
        expected_min_text_chars: int = 100,
        expected_selectors: Optional[List[str]] = None,
    ) -> QualityReport:
        reasons: List[str] = []
        score = 1.0
        completeness = 1.0
        blocked = False
        likely_unrendered = False
        suggested_escalation: Optional[str] = None

        # 1. HTTP Status sanity
        if status_code >= 400:
            score -= 0.6
            completeness = 0.0
            reasons.append(f"HTTP error status: {status_code}")
            if status_code in (403, 429):
                blocked = True
                reasons.append("Rate limit or access forbidden")
                suggested_escalation = "chromium"

        # 2. Content Type Check
        content_type = headers.get("content-type", "").lower()
        if "text/html" not in content_type and "application/xhtml" not in content_type:
            if "application/json" in content_type:
                # Could be structured API result or JSON error
                pass

        # 3. Block page detection
        text_sample = html_or_text[:10000]
        for pattern in self.BLOCK_PATTERNS:
            if pattern.search(text_sample):
                blocked = True
                score -= 0.5
                reasons.append(f"Block/Challenge pattern detected: {pattern.pattern}")
                suggested_escalation = "chromium"
                break

        # 4. Unrendered SPA shell detection
        for pattern in self.UNRENDERED_SPA_PATTERNS:
            if pattern.search(text_sample):
                likely_unrendered = True
                score -= 0.4
                completeness = min(completeness, 0.3)
                reasons.append("Empty SPA root or disabled JS message detected")
                if not suggested_escalation:
                    suggested_escalation = "servo"
                break

        # 5. Useful text content evaluation
        # Strip HTML tags
        clean_text = re.sub(r"<[^>]+>", " ", html_or_text)
        clean_text = re.sub(r"\s+", " ", clean_text).strip()
        text_len = len(clean_text)

        if text_len < expected_min_text_chars:
            score -= 0.3
            completeness = min(
                completeness, max(0.1, text_len / max(expected_min_text_chars, 1))
            )
            reasons.append(
                f"Low useful text characters: {text_len} < {expected_min_text_chars}"
            )
            if not suggested_escalation and not blocked:
                suggested_escalation = "servo"

        # 6. Expected selectors check
        if expected_selectors:
            missing = [sel for sel in expected_selectors if sel not in html_or_text]
            if missing:
                completeness -= 0.2 * (len(missing) / len(expected_selectors))
                reasons.append(f"Missing expected selectors: {missing}")

        final_score = max(0.0, min(1.0, score))
        final_completeness = max(0.0, min(1.0, completeness))

        return QualityReport(
            score=round(final_score, 3),
            completeness=round(final_completeness, 3),
            blocked=blocked,
            likely_unrendered=likely_unrendered,
            reasons=reasons,
            suggested_escalation=suggested_escalation,
        )
